nixos/modules/config/users-groups.nix

```
@@ -555,31 +558,35 @@ in {
       };
     };
 
-    users.groups = {
-      root.gid = ids.gids.root;
-      wheel.gid = ids.gids.wheel;
-      disk.gid = ids.gids.disk;
-      kmem.gid = ids.gids.kmem;
-      tty.gid = ids.gids.tty;
-      floppy.gid = ids.gids.floppy;
-      uucp.gid = ids.gids.uucp;
-      lp.gid = ids.gids.lp;
-      cdrom.gid = ids.gids.cdrom;
-      tape.gid = ids.gids.tape;
-      audio.gid = ids.gids.audio;
-      video.gid = ids.gids.video;
-      dialout.gid = ids.gids.dialout;
-      nogroup.gid = ids.gids.nogroup;
-      users.gid = ids.gids.users;
-      nixbld.gid = ids.gids.nixbld;
-      utmp.gid = ids.gids.utmp;
-      adm.gid = ids.gids.adm;
-      input.gid = ids.gids.input;
-      kvm.gid = ids.gids.kvm;
-      render.gid = ids.gids.render;
-      sgx.gid = ids.gids.sgx;
-      shadow.gid = ids.gids.shadow;
-    };
+    users.groups = mkMerge [
+      {
+        root.gid = ids.gids.root;
+        wheel.gid = ids.gids.wheel;
+        disk.gid = ids.gids.disk;
+        kmem.gid = ids.gids.kmem;
+        tty.gid = ids.gids.tty;
+        floppy.gid = ids.gids.floppy;
+        uucp.gid = ids.gids.uucp;
+        lp.gid = ids.gids.lp;
+        cdrom.gid = ids.gids.cdrom;
+        tape.gid = ids.gids.tape;
+        audio.gid = ids.gids.audio;
+        video.gid = ids.gids.video;
+        dialout.gid = ids.gids.dialout;
+        nogroup.gid = ids.gids.nogroup;
+        users.gid = ids.gids.users;
+        nixbld.gid = ids.gids.nixbld;
+        utmp.gid = ids.gids.utmp;
+        adm.gid = ids.gids.adm;
+        input.gid = ids.gids.input;
+        kvm.gid = ids.gids.kvm;
+        render.gid = ids.gids.render;
+        sgx.gid = ids.gids.sgx;
+        shadow.gid = ids.gids.shadow;
+      }
+      # Create per-user primary groups
+      (lib.mapAttrs (n: v: {}) nixosConfig.users.users)
```

This would create unused groups when a different explicit group name is assigned to a user.

If we want to implicitly create all groups referenced by users, we should
```suggestion
      (lib.mapAttrs' (_: user: { name = user.groups; value = {}; }) nixosConfig.users.users)
```
But this also has the problem that typos cannot be caught.
```nix
{
  users.users.someuser.group = "someusrr"; # Oops. Implicitly created another group
  users.groups.someuser = { /* options... */ };
}
```

Alternatively, we can force user to define the same-name groups. But this breaks existing settings and requires migration.

I'm not sure which approach is better though.

---

> This would create unused groups when a different explicit group name is assigned to a user.

Oops! I added your fix + `s/user.groups/user.group/`, which looked like a typo.

> But this also has the problem that typos cannot be caught.
> {
>   users.users.someuser.group = "someusrr"; # Oops. Implicitly created another group
>   users.groups.someuser = { /* options... */ };
> }

I can live with that.

---

> > But this also has the problem that typos cannot be caught.
> > {
> > users.users.someuser.group = "someusrr"; # Oops. Implicitly created another group
> > users.groups.someuser = { /* options... */ };
> > }
> 
> I can live with that.

Currently we can rule this error out, but this would make it less checked and more fragile. I think it's better to only create implicit groups for users whose `group` is not defined explicitly. But I'm not sure if it is doable.